sdk/java/pulumi/src/main/java/com/pulumi/provider/internal/properties/PropertyValueSerializer.java

```diff
@@ -0,0 +1,359 @@
+package com.pulumi.provider.internal.properties;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import com.pulumi.core.Output;
+import com.pulumi.core.annotations.Export;
+import com.pulumi.core.annotations.Import;
+import com.pulumi.core.internal.OutputData;
+import com.pulumi.core.internal.OutputInternal;
+import com.pulumi.resources.DependencyResource;
+import com.pulumi.resources.Resource;
+
+import com.google.common.collect.ImmutableSet;
+
+/**
+ * A utility class for deserializing {@link PropertyValue} objects into Java types.
+ * This class handles deserialization of primitive types, collections, maps, and complex objects
+ * with support for Pulumi's {@link Output} types and annotations.
+ */
+public final class PropertyValueSerializer {
+    private PropertyValueSerializer() {}
+
+    /**
+     * Deserializes a PropertyValue into the specified target type.
+     *
+     * @param value the PropertyValue to deserialize
+     * @param targetType the Class to deserialize into
+     * @param <T> the type to deserialize to
+     * @return the deserialized object of type T
+     * @throws PropertyDeserializationException if the value cannot be deserialized to the target type
+     * @throws IllegalArgumentException if value or targetType is null
+     */
+    public static <T> T deserialize(PropertyValue value, Class<T> targetType) throws PropertyDeserializationException {
+        if (targetType == null) {
+            throw new IllegalArgumentException("Target type cannot be null");
+        }
+        if (value == null) {
+            throw new IllegalArgumentException("Value cannot be null");
+        }
+
+        Object deserialized = deserializeValue(value, targetType);
+        
+        if (targetType.isInstance(deserialized)) {
+            return targetType.cast(deserialized);
+        }
+
+        throw new PropertyDeserializationException(
+            "Could not deserialize value", 
+            new String[]{"$"}, 
+            targetType, 
+            value,
+            null
+        );
+    }
+
+    private static Object deserializeValue(PropertyValue value, java.lang.reflect.Type targetType) {
+        String[] rootPath = new String[]{"$"};
+        return deserializeValue(value, targetType, rootPath);
+    }
+
+    private static Object deserializeValue(PropertyValue value, java.lang.reflect.Type targetType, String[] path) {
+        Class<?> rawType = resolveGenericType(targetType);
+        System.out.printf("deserializeValue: value.type=%s, targetType=%s, rawType=%s%n", 
+            value.getType(), targetType.getTypeName(), rawType.getName());
```

Comment:
debugging leftovers?  Should this be logged instead?

Reply:
yeah, thank you for catching!